src/core/trulens/core/database/sqlalchemy.py

```diff
@@ -715,6 +717,129 @@ def get_records_and_feedback(
+
+        # TODO: thread safety
+        with self.session.begin() as session:
+            if (
+                _ground_truth := session.query(self.orm.GroundTruth)
+                .filter_by(ground_truth_id=ground_truth.ground_truth_id)
+                .first()
+            ):
+                _ground_truth.ground_truth_json = ground_truth.model_dump_json()
+            else:
+                _ground_truth = self.orm.GroundTruth.parse(
+                    ground_truth, redact_keys=self.redact_keys
+                )
+
+                session.merge(_ground_truth)
+
+            logger.info(
+                f"{UNICODE_CHECK} added ground truth {_ground_truth.ground_truth_id}"
+            )
+
+            return _ground_truth.ground_truth_id
+
+    def batch_insert_ground_truth(
+        self, ground_truths: List[mod_groundtruth_schema.GroundTruth]
+    ) -> List[mod_types_schema.GroundTruthID]:
+        """See [DB.batch_insert_ground_truth][trulens_eval.database.base.DB.batch_insert_ground_truth]."""
+        with self.session.begin() as session:
+            ground_truths_list = []
+
+            for ground_truth in ground_truths:
+                if (
+                    _ground_truth := session.query(self.orm.GroundTruth)
+                    .filter_by(ground_truth_id=ground_truth.ground_truth_id)
+                    .first()
+                ):
+                    _ground_truth.ground_truth_json = (
+                        ground_truth.model_dump_json()
+                    )
+                else:
+                    _ground_truth = self.orm.GroundTruth.parse(
+                        ground_truth, redact_keys=self.redact_keys
+                    )
+                    ground_truths_list.append(_ground_truth)
+
+            session.bulk_save_objects(ground_truths_list)
+            return [gt.ground_truth_id for gt in ground_truths_list]
+
+    def get_ground_truth(
+        self, ground_truth_id: str | None = None
+    ) -> Optional[JSONized]:
+        """See [DB.get_ground_truth][trulens.core.database.base.DB.get_ground_truth]."""
+
+        with self.session.begin() as session:
+            if (
+                _ground_truth := session.query(self.orm.GroundTruth)
+                .filter_by(ground_truth_id=ground_truth_id)
+                .first()
+            ):
+                return json.loads(_ground_truth)
+
+    def get_ground_truths_by_dataset(
+        self, dataset_name: str
+    ) -> pd.DataFrame | None:
+        """See [DB.get_ground_truths_by_dataset][trulens.core.database.base.DB.get_ground_truths_by_dataset]."""
+        with self.session.begin() as session:
+            q = sa.select(self.orm.Dataset)
+            all_datasets = (row[0] for row in session.execute(q))
+            df = None
+            for dataset in all_datasets:
+                dataset_json = json.loads(dataset.dataset_json)
+                if (
```

Comment:
Can there be multiple datasets with the same name?

Reply:
Yes if the metadata is different. 

The dataset id is a hash based on its content (dataset_json + name). This ensures idempotency + typical use cases where people running experiments might have the same dataset containing multiple different subsets (think of TREC datsets)